### PR TITLE
test: add tests for longestcommonsubsequence.go

### PR DESCRIPTION
**Pull request by vil02**
#### Description of Change

I added tests for [`longestcommonsubsequence.go`](https://github.com/TheAlgorithms/Go/blob/51f0485f9289676e23d4c7cbf4a868e4704a27d6/dynamic/longestcommonsubsequence.go). There are two new tests: the first one checks is a basic check of the result of the function `LongestCommonSubsequence`. The second one checks if the function `LongestCommonSubsequence` is _symmetric_.
The example data was verified using:
- https://www.geeksforgeeks.org/longest-common-subsequence-dp-4/
- https://github.com/TheAlgorithms/Python/blob/master/dynamic_programming/longest_common_subsequence.py

#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/Go/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [GoDoc guidelines](https://blog.golang.org/godoc)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/Go/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: 
Added tests for [`longestcommonsubsequence.go`](https://github.com/TheAlgorithms/Go/blob/51f0485f9289676e23d4c7cbf4a868e4704a27d6/dynamic/longestcommonsubsequence.go).

### CODE REVIEW

**Comment by vil02** [comment]
The new changes are analogous to the ones suggested by @tjgurwara99 in the PR #554 and #555.

File: dynamic/longestcommonsubsequence_test.go
Comment (tjgurwara99) [inline dynamic/longestcommonsubsequence_test.go]:
```suggestion
type testCases struct {
```

File: dynamic/longestcommonsubsequence_test.go
Comment (tjgurwara99) [inline dynamic/longestcommonsubsequence_test.go]:
```suggestion
func getTestCases() []LongestCommonSubsequenceExample {
```

**Review by tjgurwara99 — CHANGES_REQUESTED** [review]

File: dynamic/longestcommonsubsequence_test.go
Comment (tjgurwara99) [inline dynamic/longestcommonsubsequence_test.go:40]:
```suggestion
				"expected to be: %d, but got: %d",
```

File: dynamic/longestcommonsubsequence_test.go
Comment (tjgurwara99) [inline dynamic/longestcommonsubsequence_test.go:53]:
```suggestion
				"expected to be: %d, but got: %d",
```

**Review by tjgurwara99 — CHANGES_REQUESTED** [review]

**Review by tjgurwara99 — CHANGES_REQUESTED** [review]

File: dynamic/longestcommonsubsequence_test.go
Comment (vil02) [reply dynamic/longestcommonsubsequence_test.go:53]: It is clear from the statement of the problem, that the function `LongestCommonSubsequence` is symmetric. Just from looking at the implementation it is not so obvious that it has such property. I think it is worth to have this test, although the same data is used. Testing it against some random input (or even fuzzing) might be an overkill in this case, in my opinion.

Comment (tjgurwara99) [reply dynamic/longestcommonsubsequence_test.go:53]: This is a **TEST** not an implementation; so your argument doesn't really make sense to me. If you're concerned about it being not clear then change the implementation or add some documentation about it, why add something that is a duplicate of the same thing while also **calling** it the same thing, i.e. a `Test`? This will run twice, make our CI test slow (by one test which shouldn't need to exist) and doesn't provide any benefit. I see no reason this should be another test.

Comment (tjgurwara99) [reply dynamic/longestcommonsubsequence_test.go:53]: Also, what you might want is an `Example` prefixed function - read more about them [here](https://go.dev/blog/examples).

Comment (vil02) [reply dynamic/longestcommonsubsequence_test.go:53]: `LongestCommonSubsequence` does not need to be implemented in such a way that the symmetry is visible (unless the symmetry is used, but in our case: it is not).

What if I would extend the array in `getLCSTestCases()` by adding the same entries but with flipped strings? Then the information about the symmetry would be lost. Imagine that someone in the future would decide to add a new test case? Would a flipped test case also be added? Why to have such `duplication`?

`TestLongestCommonSubsequenceIsSymmetric` adds structure. It also makes the _requierment_ that `LongestCommonSubsequence` is symmetric _verifiable_. A comment in the documentation that `LongestCommonSubsequence` is symmetric is not verifiable - such comments are not checked by CI or any other tool.

I do not agree with the sentence about the CI taking too long. One could ask _why do we need CI at all if our code has no errors?_

Comment (tjgurwara99) [reply dynamic/longestcommonsubsequence_test.go:53]: Ok, I guess there are two things that I've misunderstood. Firstly I thought the line 36 and 47 were the same with *no* difference; which made me think that the you were running the *exact* test twice, hence my resistance against having it. Second thing I misunderstood is your intent (which is because of the first misunderstanding) behind the test.

Though I still stand by the argument that `LongestCommonSubsequence` is already symmetric by nature, therefore I don't think we need to verify it after it has been **tested** to be correct in the first place. Because otherwise (lets consider for the sake of argument that) the longest subsequence of string A and string B is X but the longest subsequence of string B and string A is Y then by (common) uniqueness argument (proof by contradiction) its quite easy to prove that X = Y. So this verification test is actually making sure that your first test was correct but that is pointless in my mind, but oh well.

Now, I think I understand what you are trying to do. Since you feel strongly about this, therefore I recommend something slightly different from what you have proposed. Its similar but I prefer the tests to be **complete** (even though I think it is complete) and since the completeness of `LongestCommonSubsequence` depends on its symmetric nature (according to you) it might be best to put it like the following:

```Go
func TestLongestCommonSubsequence(t *testing.T) {
	t.Run("Simple test", func(t *testing.T) {
		for _, tc := range getLCSTestCases() {
			actual := dynamic.LongestCommonSubsequence(
				tc.stringA, tc.stringB,
	 			len(tc.stringA), len(tc.stringB))
			if actual != tc.expected {
				t.Errorf("expected to be: %d, but got: %d", tc.expected, actual)
			}
	 	}
	})

	t.Run("Symmetric test", func(t *testing.T) {
 		for _, tc := range getLCSTestCases() {
 			actual := dynamic.LongestCommonSubsequence(
 				tc.stringB, tc.stringA,
	 			len(tc.stringB), len(tc.stringA))
 			if actual != tc.expected {
 				t.Errorf("expected to be: %d, but got: %d", tc.expected, actual)
 			}
	 	}
	})
 }

```

Comment (tjgurwara99) [reply dynamic/longestcommonsubsequence_test.go:53]: > I do not agree with the sentence about the CI taking too long. One could ask why do we need CI at all if our code has no errors?

Everything in software has tradeoffs. For me ensuring that the costs (carbon emissions of actions running) of CI are low is a valid concern (to me). I'm sure you will come across situations in your career that will answer the question of _why do we need CI at all if our code has no errors__, so I won't get too much in detail there 😂

**Review by tjgurwara99 — APPROVED** [review]
All looks good to me, thanks!

**Comment by tjgurwara99** [comment]
@vil02 just a note: once this gets merged I don't think you will need an approval to run workflows in the other PR's

**Review by raklaptudirm — APPROVED** [review]

**Pull request merged**
